src/interfaces/IMiddlewareOptions.ts

```diff
@@ -1,11 +1,27 @@
 import { ClassType } from "class-transformer-validator";
 
-export interface IMiddlewareOptions<T extends object> {
+export interface IDeprecatedMiddlewareOptions<T extends object> {
   classType: ClassType<T>;
 }
 
-export function isMiddlewareOptions(
-  options: any,
-): options is IMiddlewareOptions<any> {
-  return !!(options && options.classType);
+export interface IBodyMiddlewareOptions<T extends object> {
+  bodyClassType: ClassType<T>;
+}
+
+export interface IQueryMiddlewareOptions<T extends object> {
+  queryClassType: ClassType<T>;
 }
+
+export type IMiddlewareOptions<T extends object> =
+  | IDeprecatedMiddlewareOptions<T>
+  | IQueryMiddlewareOptions<T>
+  | IBodyMiddlewareOptions<T>;
+
+export const isMiddlewareOptions = (
+  options: any,
+): options is IMiddlewareOptions<any> =>
+  !!(
+    options &&
+    (options.classType !== options.bodyClassType ||
+      (options.queryClassType && !(options.classType && options.bodyClassType)))
+  );
```

Comment:
This is becoming really hard to read, could you maybe extract into something like
```suggestion
    isDeprecatedMiddlewareOptions(options)
    || isBodyMiddlewareOptions(options)
    || isQueryMiddlewareOptions(options)
    || isBothMiddlewareOptions(options)
```